### PR TITLE
Second order swerve fix

### PR DESCRIPTION
**Pull request by josieammer**
Compensate for lateral drift while rotating.  Adds a discretize function to DrivetrainSubsystem for correcting chassisSpeeds.  Time scale factor found empirically via testing on robot driving a straight line forward for ~10ft with a 25% rotation, and measuring the lateral offset.  Remaining error is approximately 1' lateral error over ~10f driving distance.

### CODE REVIEW

File: smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java [hunk 1]
```diff
@@ -616,6 +615,25 @@ private ChassisSpeeds clampChassisSpeeds(ChassisSpeeds chassisSpeeds){
...
+  private ChassisSpeeds discretize(ChassisSpeeds speeds) {
+    // a fudge factor to increase the size of the discretization correction. 
+    // other teams use [1..4]
+    double timeScaleFactor = 1.9; 
+    var desiredDeltaPose = new Pose2d(
+      speeds.vxMetersPerSecond * deltaTimeSeconds, 
+      speeds.vyMetersPerSecond * deltaTimeSeconds, 
+      new Rotation2d(speeds.omegaRadiansPerSecond * deltaTimeSeconds * timeScaleFactor)
+    );
+    var twist = new Pose2d().log(desiredDeltaPose);
```
Comment (mmcadamsms) [inline smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:627]: so this is the magic?

image

File: smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java [hunk 2]
```diff
@@ -616,6 +615,25 @@ private ChassisSpeeds clampChassisSpeeds(ChassisSpeeds chassisSpeeds){
...
+
+    return new ChassisSpeeds(
+      (twist.dx / deltaTimeSeconds),
```
Comment (mmcadamsms) [inline smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:630]: I also read ... https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/math/geometry/Twist2d.html 

image

Comment (mmcadamsms) [reply smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:630]: seems this turns it back into a velocity ... is that correct?

**Review by mmcadamsms — APPROVED** [review]

File: smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java (hunk 1 above)
Comment (josieammer) [reply smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:627]: Yep!

File: smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java (hunk 2 above)
Comment (josieammer) [reply smokey/src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:630]: Yep.

**Pull request merged**